src/pyproject_hooks/_in_process/_in_process.py

```diff
@@ -59,38 +56,51 @@ def __init__(self, hook_name=None):
         self.hook_name = hook_name
 
 
-def contained_in(filename, directory):
-    """Test if a file is located within the given directory."""
-    filename = os.path.normcase(os.path.abspath(filename))
-    directory = os.path.normcase(os.path.abspath(directory))
-    return os.path.commonprefix([filename, directory]) == directory
-
-
 def _build_backend():
     """Find and load the build backend"""
     # Add in-tree backend directories to the front of sys.path.
     backend_path = os.environ.get("PEP517_BACKEND_PATH")
-    if backend_path:
-        extra_pathitems = backend_path.split(os.pathsep)
-        sys.path[:0] = extra_pathitems
-
     ep = os.environ["PEP517_BUILD_BACKEND"]
     mod_path, _, obj_path = ep.partition(":")
-    try:
-        obj = import_module(mod_path)
-    except ImportError:
-        raise BackendUnavailable(traceback.format_exc())
 
     if backend_path:
-        if not any(contained_in(obj.__file__, path) for path in extra_pathitems):
-            raise BackendInvalid("Backend was not loaded from backend-path")
+        extra_pathitems = backend_path.split(os.pathsep)
+        obj = _load_module_from_path(mod_path, extra_pathitems)
+    else:
+        try:
+            obj = import_module(mod_path)
+        except ImportError:
+            msg = f"Cannot import {mod_path!r}"
+            raise BackendUnavailable(msg, traceback.format_exc())
 
     if obj_path:
         for path_part in obj_path.split("."):
             obj = getattr(obj, path_part)
     return obj
 
 
+def _load_module_from_path(fullname, pathitems):
+    """Given a set of sys.path-like entries, load a module from it"""
+    sys.path[:0] = pathitems  # Still required for other imports.
+    parts = fullname.split(".")
+    # Parent packages need to be imported to ensure everything comes from pathitems.
+    for i in range(len(parts)):
+        module_name = ".".join(parts[: i + 1])
+        spec = _find_spec_in_path(module_name, pathitems)
+        module = module_from_spec(spec)
+        sys.modules[module_name] = module
+        spec.loader.exec_module(module)
```

Comment:
This will execute the module with `sys.meta_path` unmodified. Even if you fixed loading this particular module from the right place, modules it imports during execution will be found by the loaders in `sys.meta_path`, which is even more dangerous than loading the specified backend module from the wrong place IMO.

> ```python
> # Parent packages need to be imported to ensure everything comes from pathitems.
> ```

Exactly, but children and packages with a different top-level name too!

IMO the correct way to handle this is to simply prepend your `PathFinder` to `sys.meta_path` while we perform the build, and just import as normal, instead of this custom import hack :sweat_smile:

Reply:
Thank you very much for the review, Filipe.

In f14e5df I tried to implement the suggested approach. Please let me know if I misunderstood your comment (I can always revert and try something different).